### PR TITLE
[deckhouse-controller] Calculate result release deploy time

### PR DESCRIPTION
**Pull request by libmonsoon-dev**
## Description
Added calculation of the nearest possible time for release deployment. Now, if there are several factors limiting deployment, all reasons will be indicated in the release status. In addition, sending metrics about blocked module releases has been removed if the corresponding module is inactive

## Why do we need it, and what problem does it solve?
Close https://github.com/deckhouse/deckhouse/issues/8990

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-controller
type: feature
summary: Now, if there are several factors limiting deployment, all reasons with the nearest possible moment of deployment will be indicated in the release status. In addition, sending metrics about blocked module releases has been removed if the corresponding module is inactive
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/openstack'...

**Comment by libmonsoon-dev** [comment]
/e2e/abort openstack;Standard;containerd;1.31 11188704902 11188704902-1-con-1-31 /sys/deckhouse-oss/install:pr9988

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/openstack'...

**Comment by libmonsoon-dev** [comment]
/e2e/abort openstack;Standard;containerd;1.31 11194055487 11194055487-1-con-1-31 /sys/deckhouse-oss/install:pr9988

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've recognized your '/e2e/abort' command and started the workflow...

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 1]
```diff
@@ -526,21 +521,33 @@ func (c *moduleReleaseReconciler) reconcilePendingRelease(ctx context.Context, m
...
+		// TODO: requeue all releases if deckhouse update settings is changed
+		// requeueAfter := notReadyErr.RetryDelay()
+		// if requeueAfter == 0 {
+		// requeueAfter = defaultCheckInterval
+		// }
+		// r.logger.Infof("%s: retry after %s", err.Error(), requeueAfter)
+		return ctrl.Result{Requeue: true, RequeueAfter: notReadyErr.RetryDelay()}, nil
+	}
+
+	return ctrl.Result{RequeueAfter: defaultCheckInterval}, fmt.Errorf("apply predicted release: %w", err)
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:545]: If the returned error is non-nil, the Result is ignored and the request will be requeued using exponential backoff. The only exception is if the error is a TerminalError in which case no requeuing happens.

File: go_lib/updater/updater.go [hunk 2]
```diff
@@ -46,6 +51,75 @@ const (
...
+	}
+
+	if len(reasons) != 0 {
+		str := strings.Join(reasons, ". ")
+		if applyTime.IsZero() {
+			return str
+		}
+
+		if r.contains(manualApprovalRequiredReason) {
+			str += ". After approval the release will be delayed"
```
Comment (ldmonster) [inline go_lib/updater/updater.go:102]: Maybe replace with StringBuilder?

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go [hunk 3]
```diff
@@ -361,18 +361,29 @@ func (r *deckhouseReleaseReconciler) pendingReleaseReconcile(ctx context.Context
...
+		// TODO: requeue all releases if deckhouse update settings is changed
+		// requeueAfter := notReadyErr.RetryDelay()
+		// if requeueAfter == 0 {
+		// requeueAfter = defaultCheckInterval
+		// }
+		// r.logger.Infof("%s: retry after %s", err.Error(), requeueAfter)
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	return ctrl.Result{RequeueAfter: defaultCheckInterval}, fmt.Errorf("apply predicted release: %w", err)
```
[the remark this answers is not in the document]
Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/deckhouse-release/controller.go:384]: this is the point, isn't it? we want to requeue the release if something went wrong when processing it

**Comment by miklezzzz** [comment]
would be cool to have some straightforward examples from a dev cluster, like :smiling_face_with_tear: :
before
```
root@dev-master-0:~# kubectl get deckhousereleases.deckhouse.io 
NAME       PHASE     TRANSITIONTIME   MESSAGE
v1.63.11   Pending   4s               Release is waiting for the update window: 08 Oct 24 08:00 UTC
```
after
```
root@dev-master-0:~# kubectl get deckhousereleases.deckhouse.io 
NAME       PHASE     TRANSITIONTIME   MESSAGE
v1.63.11   Pending   100s             Release is postponed by canary process. Release is waiting for the update window until 12 Oct 24 08:00 UTC
```

```
apiVersion: deckhouse.io/v1alpha1
approved: false
kind: DeckhouseRelease
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"deckhouse.io/v1alpha1","approved":false,"kind":"DeckhouseRelease","metadata":{"annotations":{},"name":"v1.63.11"},"spec":{"applyAfter":"2024-10-10T10:48:00Z","changelog":{"candi":{"fixes":[{"pull_request":"https://github.com/deckhouse/deckhouse/pull/9960","summary":"fix resize partition step"}]}},"changelogLink":"https://github.com/deckhouse/deckhouse/releases/tag/v1.63.11","version":"v1.63.11"}}
    release.deckhouse.io/isUpdating: "false"
    release.deckhouse.io/notified: "true"
  creationTimestamp: "2024-10-07T10:18:35Z"
  generation: 1
  name: v1.63.11
  resourceVersion: "489766627"
  uid: 55016419-ed10-4723-82cb-1f44f40f0263
spec:
  applyAfter: "2024-10-10T10:48:00Z"
  changelog:
    candi:
      fixes:
      - pull_request: https://github.com/deckhouse/deckhouse/pull/9960
        summary: fix resize partition step
  changelogLink: https://github.com/deckhouse/deckhouse/releases/tag/v1.63.11
  version: v1.63.11
status:
  approved: false
  message: Release is postponed by canary process. Release is waiting for the update
    window until 12 Oct 24 08:00 UTC
  phase: Pending
  transitionTime: "2024-10-07T10:22:23Z"
  ```

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/deckhouse-release/controller.go:384]: fixed

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 4]
```diff
@@ -1166,22 +1174,22 @@ func (c *moduleReleaseReconciler) cleanUpModuleReleases(ctx context.Context, mr
...
 					return ctrl.Result{Requeue: true}, fmt.Errorf("couldn't clean up outdated release %q of %s module: %w", releases[i].name, moduleName, err)
 				}
-				c.logger.Infof("cleaned up outdated release %q of %q module", releases[i].name, moduleName)
+				r.logger.Infof("cleaned up outdated release %q of %q module", releases[i].name, moduleName)
 			}
 		}
 	}
 
-	return ctrl.Result{}, nil
+	return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
```
[the remark this answers is not in the document]
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go]: fixed

File: go_lib/updater/updater.go (hunk 2 above)
Comment (libmonsoon-dev) [reply go_lib/updater/updater.go]: fixed

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go [hunk 5]
```diff
@@ -211,29 +214,35 @@ func (k *kubeAPI) SaveReleaseData(ctx context.Context, release *v1alpha1.ModuleR
...
 	}
 }
 
 func (m *metricsUpdater) ReleaseBlocked(_, _ string) {
-
 }
 
-func (m *metricsUpdater) WaitingManual(name string, _ float64) {
-	m.metricStorage.GaugeSet("d8_module_release_waiting_manual", 1, map[string]string{"name": name})
+func (m *metricsUpdater) WaitingManual(name string, totalPendingManualReleases float64) {
```
Comment (miklezzzz) [inline deckhouse-controller/pkg/controller/module-controllers/release/updater.go:240]: `name` argument contains the module release name, not the module name, we can't use it in the following check.

Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go:240]: would it be difficult to add some tests for metrics? quite annoying to test it manually :\

File: go_lib/updater/updater.go [hunk 6]
```diff
@@ -46,6 +51,75 @@ const (
...
+	notificationDelayReason
+	outOfWindowReason
+	manualApprovalRequiredReason
+)
+
+func (r deployDelayReason) String() string {
+	return r.string(time.Time{})
+}
+
+func (r deployDelayReason) string(applyTime time.Time) string {
```
[the remark this answers is not in the document]
Comment (libmonsoon-dev) [reply go_lib/updater/updater.go:69]: Done

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go (hunk 5 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go]: There were too many changes regarding the metrics, I will make a separate pull request for this. In this PR I rolled back the corresponding changes

**Review by miklezzzz — APPROVED** [review]

**Review by ldmonster — APPROVED** [review]

**Review by yalosev — APPROVED** [review]
Conflicts, please

**Pull request merged**